### PR TITLE
[receiver/azureblob] Support service principal authentication for Azure Blob storage

### PR DESCRIPTION
**Pull request by sebastiean**
**Description:**
Added support for authenticating to Blob storage via a Service Principal.
The default authentication method is still connection string to ensure backwards compatibility.

**Link to tracking Issue:** #32705 

**Testing:** 
Extended the unit tests for the blob client and the config

**Documentation:** 
Added information and examples to README

### CODE REVIEW

**Review by mx-psi — COMMENTED** [review]
Looks like there are some CI failures (you probably need to `make gotidy`?). Can you take a look?

**Comment by sebastiean** [comment]
> Looks like there are some CI failures (you probably need to `make gotidy`?). Can you take a look?

I'll take a look!

File: receiver/azureblobreceiver/README.md [hunk 1]
```diff
@@ -19,25 +19,55 @@ This receiver reads logs and trace data from [Azure Blob Storage](https://azure.
...
 The following settings are required:
 
-- `connection_string:` (no default): Azure Blob Storage connection key, which can be found in the Azure Blob Storage resource on the Azure Portal.
 - `event_hub:`
   `  endpoint:` (no default): Azure Event Hub endpoint triggering on the `Blob Create` event 
 
 The following settings can be optionally configured:
 
+- `auth` (default = connection_string): Specifies the used authentication method. Supported values are `connection_string`, `service_principal`.
+- `cloud` (default = "AzureCloud"): defines which Azure cloud to use when using the `service_principal` authentication method. Either `AzureCloud` or `AzureUSGovernment`
```
Comment (tiffany76) [inline receiver/azureblobreceiver/README.md:28]:
```suggestion
- `cloud` (default = "AzureCloud"): Defines which Azure Cloud to use when using the `service_principal` authentication method. Value is either `AzureCloud` or `AzureUSGovernment`.
```

File: receiver/azureblobreceiver/README.md [hunk 2]
```diff
@@ -19,25 +19,55 @@ This receiver reads logs and trace data from [Azure Blob Storage](https://azure.
...
 - `logs:`
   `  container_name:` (default = "logs"): Name of the blob container with the logs
 - `traces:`
   `  container_name:` (default = "traces"): Name of the blob container with the traces
 
-Example:
+Authenticating using a connection string requires following additional settings:
```
Comment (tiffany76) [inline receiver/azureblobreceiver/README.md:34]:
```suggestion
Authenticating using a connection string requires configuration of the following additional setting:
```

File: receiver/azureblobreceiver/README.md [hunk 3]
```diff
@@ -19,25 +19,55 @@ This receiver reads logs and trace data from [Azure Blob Storage](https://azure.
...
+
+- `connection_string:` Azure Blob Storage connection key, which can be found in the Azure Blob Storage resource on the Azure Portal.
+
+Authenticating using service principal requires following additional settings:
```
Comment (tiffany76) [inline receiver/azureblobreceiver/README.md:38]:
```suggestion
Authenticating using service principal requires configuration of the following additional settings:
```

File: receiver/azureblobreceiver/README.md [hunk 4]
```diff
@@ -19,25 +19,55 @@ This receiver reads logs and trace data from [Azure Blob Storage](https://azure.
...
+
+- `tenant_id`
+- `client_id`
+- `client_secret`
+- `storage_account_url`
+
+The service principal requires the [Storage Blob Data Contributor](https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles/storage#storage-blob-data-contributor) role on the logs and traces containers.
```
Comment (tiffany76) [inline receiver/azureblobreceiver/README.md:45]:
```suggestion
The service principal method also requires the [Storage Blob Data Contributor](https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles/storage#storage-blob-data-contributor) role on the logs and traces containers.
```

File: receiver/azureblobreceiver/README.md [hunk 5]
```diff
@@ -19,25 +19,55 @@ This receiver reads logs and trace data from [Azure Blob Storage](https://azure.
...
+
+### Example Configurations
```
Comment (tiffany76) [inline receiver/azureblobreceiver/README.md:47]:
```suggestion
### Example configurations
```

File: receiver/azureblobreceiver/README.md [hunk 6]
```diff
@@ -19,25 +19,55 @@ This receiver reads logs and trace data from [Azure Blob Storage](https://azure.
...
 ```yaml
 receivers:
   azureblob:
     connection_string: DefaultEndpointsProtocol=https;AccountName=accountName;AccountKey=+idLkHYcL0MUWIKYHm2j4Q==;EndpointSuffix=core.windows.net
     event_hub:
-      endpoint: Endpoint=sb://oteldata.servicebus.windows.net/;SharedAccessKeyName=otelhubbpollicy;SharedAccessKey=mPJVubIK5dJ6mLfZo1ucsdkLysLSQ6N7kddvsIcmoEs=;EntityPath=otellhub    
+      endpoint: Endpoint=sb://oteldata.servicebus.windows.net/;SharedAccessKeyName=otelhubbpollicy;SharedAccessKey=mPJVubIK5dJ6mLfZo1ucsdkLysLSQ6N7kddvsIcmoEs=;EntityPath=otellhub
+```
+
+Using Service Principal for authentication:
```
Comment (tiffany76) [inline receiver/azureblobreceiver/README.md:59]:
```suggestion
Using service principal for authentication:
```

**Review by tiffany76 — COMMENTED** [review]
Just a few copy-edit suggestions from a docs perspective. Thanks!

**Comment by sebastiean** [comment]
> Just a few copy-edit suggestions from a docs perspective. Thanks!

Thanks @tiffany76. I've committed those ✅

File: receiver/azureblobreceiver/config.go [hunk 7]
```diff
@@ -3,10 +3,44 @@
...
+	// Storage Account URL, used with Service Principal authentication
+	StorageAccountURL string `mapstructure:"storage_account_url"`
+	// Tenant ID, used with Service Principal authentication
+	TenantID string `mapstructure:"tenant_id"`
+	// Client ID, used with Service Principal authentication
+	ClientID string `mapstructure:"client_id"`
+	// Client secret, used with Service Principal authentication
+	ClientSecret string `mapstructure:"client_secret"`
+	// Azure Cloud to authenticate against, used with Service Principal authentication
+	Cloud string `mapstructure:"cloud"`
```
Comment (mx-psi) [inline receiver/azureblobreceiver/config.go:43]: nit: you can use an enum type here like we do here: https://github.com/open-telemetry/opentelemetry-collector/blob/8b5a656dce445de4036a7b72ff7d76bb74298572/config/configcompression/compressiontype.go#L27-L41 if this has a fixed set of values

File: receiver/azureblobreceiver/config.go [hunk 8]
```diff
@@ -3,10 +3,44 @@
...
 	// which can be found in the Azure Blob Storage resource on the Azure Portal. (no default)
 	ConnectionString string `mapstructure:"connection_string"`
+	// Storage Account URL, used with Service Principal authentication
+	StorageAccountURL string `mapstructure:"storage_account_url"`
+	// Tenant ID, used with Service Principal authentication
+	TenantID string `mapstructure:"tenant_id"`
+	// Client ID, used with Service Principal authentication
+	ClientID string `mapstructure:"client_id"`
+	// Client secret, used with Service Principal authentication
+	ClientSecret string `mapstructure:"client_secret"`
```
Comment (mx-psi) [inline receiver/azureblobreceiver/config.go:41]: Sensitive configuration fields should use `configopaque.String`
```suggestion
	ClientSecret configopaque.String `mapstructure:"client_secret"`
```

File: receiver/azureblobreceiver/config.go [hunk 9]
```diff
@@ -3,10 +3,44 @@
...
+	errMissingClientID          = errors.New(`"ClientID" is not specified in config`)
+	errMissingClientSecret      = errors.New(`"ClientSecret" is not specified in config`)
+	errMissingStorageAccountURL = errors.New(`"StorageAccountURL" is not specified in config`)
+	errMissingConnectionString  = errors.New(`"ConnectionString" is not specified in config`)
+	errInvalidCloud             = errors.New(`"Cloud" is invalid`)
+)
+
 type Config struct {
+	// Type of authentication to use
+	Authentication string `mapstructure:"auth"`
```
Comment (mx-psi) [inline receiver/azureblobreceiver/config.go:30]: You can also use an enum here

File: receiver/azureblobreceiver/config.go (hunk 7 above)
Comment (mx-psi) [inline receiver/azureblobreceiver/config.go:43]: Would it make sense to have a `service_principal` section for all of these settings?

**Comment by sebastiean** [comment]
Thanks @mx-psi 
I've made the suggested changes

**Review by mx-psi — APPROVED** [review]

**Pull request merged**